networkx/algorithms/bipartite/matching.py

```diff
@@ -587,3 +596,363 @@ def minimum_weight_full_matching(G, top_nodes=None, weight="weight"):
+            X_i -> neighbors of Y_i that were matched by M
+
+    The iterative pattern stops when either Y_k or X_k are an empty set for a
+    certain k, forever staying an empty set. From k onward the alternating sequence
+    is proven to always stop in the article.
+
+    Two tuples containing sets of nodes are returned - tuple(X_subsets), tuple(Y_subsets).
+
+    >>> G = nx.complete_bipartite_graph(3,3)
+    >>> M = nx.bipartite.hopcroft_karp_matching(G)
+    >>> _M_alternating_sequence(G, M)
+    ((), ())
+
+    >>> G = nx.Graph([(0,3),(0,4),(1,4),(2,4)])
+    >>> M = {0:3, 3:0, 1:4, 4:1}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}), ({4},))
+
+    >>> G = nx.Graph([(0,3),(1,3),(1,4),(2,4)])
+    >>> M = {0:3,3:0,4:1,1:4}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}, {0}), ({4}, {3}))
+
+    >>> G = nx.Graph([(0,3),(1,3),(1,4),(2,4)])
+    >>> M = {0:3,3:0,4:1,1:4}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}, {0}), ({4}, {3}))
+
+    >>> G = nx.Graph([(0,6),(1,6),(1,7),(2,6),(2,8),(3,9),(3,6),(4,8),(4,7),(5,9)])
+    >>> M = {0:6,6:0,1:7,7:1,2:8,8:2,3:9,9:3}
+    >>> _M_alternating_sequence(G, M)
+    (({4, 5}, {1, 2, 3}, {0}), ({8, 9, 7}, {6}))
+
+    """
+    X, Y = nx.bipartite.sets(G, top_nodes=top_nodes)
+    X_0 = set(X) - set(M.keys())
+    if len(X_0) == 0:
+        return (), ()
+
+    X_subsets = [X_0]
+    Y_subsets = []
+
+    Y_subgroups_accumulate = set()
+
+    while len(X_subsets) == 1 or (len(X_subsets[-1]) != 0 and len(Y_subsets[-1]) != 0):
+        Y_current = {
+            nbr
+            for n in X_subsets[-1]
+            for nbr in G[n]
+            if nbr not in Y_subgroups_accumulate
+            if M.get(n) != nbr
+            if M.get(nbr) != n
+        }
+        if len(Y_current) == 0:
+            break
+        X_current = {
+            nbr
+            for n in Y_current
+            for nbr in G[n]
+            if M.get(n) == nbr and M.get(nbr) == n
+        }
+        if len(X_current) == 0:
+            break
+
+        Y_subsets.append(Y_current)
+        X_subsets.append(X_current)
+
+        Y_subgroups_accumulate.update(Y_current)
+
+    return tuple(X_subsets), tuple(Y_subsets)
+
+
+def envy_free_matching_partition(G, *, M=None, top_nodes=None):
+    r"""Return a unique EFM (Envy Free Matching) partition of bipartite graph `G`.
+
+    An EFM partition of a bipartite graph with bipartite sets X,Y is a unique partition of X - X_L, X_S and Y - X_L, X_S.
+    This unique partition is proven to have the following properties:
+    1. Every matching that matches all the nodes in X_L and is contained in G[X_L,Y_L] (all edges are between X_L and Y_L)
+    is an envy-free matching.
+    2. Every envy-free matching in `G` is contained in G[X_L,Y_L].
+
+    Due to these properties we can say that X_L and Y_L are the "*good*" nodes and X_S and Y_S are the "*bad*" nodes,
+    because - in the context of finding envy-free matchings, we will only find them in X_L and Y_L.
```

Comment:
There is a bunch of notation here that isn't introduced or explained.
Why are the subscripts `L` and `S`? That must stand for something.
What is the notation `X - X_L, X_S` representing? It doesn't look like a partition of X unless X_L is X_S. Or maybe it is a partial partition? or subtraction of sets is not what I'm expecting.

What is G[X_L, Y_L]?  

A good place to start would probably be to explain very briefly what an envy free matching is.
Make the first sentence shorter (it is used for links to the documentation for this function. Qualify terms like `G` being bipartite in the first sentence of the paragraph rather than the first line of the doc_string. 
Also, you can use latex in the paragraph -- but keep it simple because that's what looks ok in the docs.
 
Here's a start, but I don't know this material well, so please correct/change.
```
    r"""Returns an envy free matching partition of `G`.

    An Envy Free Matching (EFM) partition of a bipartite graph `G`
    with bipartite sets $(X, Y)$ is a unique partition into $(X - X_L, X_S)$
    and $(Y - X_L, X_S)$ where subscripts `L` and `S` denote....
```

Reply:
Can you give me some feedback on this change of docs? Or make the changes in the PR directly?
Thanks!

Reply:
Do you prefer that I'll explain the unclear things you mentioned here in the conversation and we can write docs together? I can also just edit the documentation in the code if you prefer that

Reply:
I'll comment here for now, I think it can help quicken this process
the sets `X_L`, `X_S` and `Y_L`, `Y_S` are disjoint sets and  are defined as a partition of `X` and `Y` respectively, i.e `X = X_L union X_S` and `Y = X_L union X_S`
Furthermore, by definition (see article) `X_L = X \ X_S` and `Y_L = Y \ Y_S`
the algorithm works by constructing `X_S` and `Y_S` following a M-alternating sequence, hence `S` means **sequence**,
`X_L` and `Y_L` are the leftover nodes which didn't make it into `X_S` and `Y_S`, hence they're called **leftover**

I wrote " An EFM partition of a bipartite graph with bipartite sets X,Y is a unique partition of X - X_L, X_S and Y - X_L, X_S."
There is a typo - it should be X - X_L, X_S and Y - Y_L, Y_S. - sorry for this mistake!
perhaps the use of the `-` character is confusing as well. maybe it can be changed to "X := X_L U X_S and Y := Y_L U Y_S".
What do you think?

`G[X_L, Y_L]` means a subgraph of G induced by the nodes in `X_L` and `Y_L` - it means you take all the nodes and edges that connect nodes from `X_L` and `Y_L`, I'll add some brief explanation of this.

Reply:
Thanks -- 

I think we can start with the definition of an Envy-Free-Matching partition:
```
An EFM partition of a bipartite graph with bipartite sets X,Y is a unique partition of X - X_L, X_S and Y - X_L, X_S.
```
That really doesn't define the EFM partition.  It says that it is a unique partition. But which one? What property separates it from all the other partitions?

Wikipedia says an [envy-free matching](https://en.wikipedia.org/wiki/Envy-free_matching) is defined as (my words):
```
Suppose G is a bipartite graph G = (X+Y, E). An envy-free matching is a
matching in which no unmatched node in X is adjacent to a matched node in Y.
That is, the matched nodes are not connected to unmatched nodes.
```

Notation:  Let's use latex in the doc_strings to write the symbols. Latex can be used in Sphinx processed doc_strings just by including the code between dollar signs.  And while the [relative complement](https://en.wikipedia.org/wiki/Complement_(set_theory)) of B within a set A is written A\B, it is also called the "set difference" and written A - B. The latter is close to Python code, so let's use that notation. Also, we might be able to avoid X_L completely by talking about "partitioning the nodes of X into those that are in the alternating sequence $X_S$ and those that are not." But I suspect you want to use $X_L$ in some places because those are the nodes that will be in the matching. So, in that sense $X_L$ is the more important set.  Maybe it should be called $X_M$ since those nodes are in the maximal matching.  But you can choose.

Perhaps the properties you say are "proven" are the properties that define the partition??? 
How about this:
```
An envy-free matching (EFM) partition of a bipartite graph $G(X+Y, E)$ separates nodes of G into those in the matching $X_M$ and $Y_M$, and those not in the matching, call them $X_S$ and $Y_S$. It is proven that $X_S$ and $Y_S$ form a M-alternating sequence for the envy-free matching $M$. The partition also represents a partition or X and of Y. It has been proven that the partition is unique for $G$, that every envy-free matching in $G$ is contained in the induced subgraph $G[X_M \union Y_M]$, and that every matching that matches all the nodes in $X_M$ and is contained in $G[X_M \union Y_M]$ is an envy-free matching.
```

I think a paragraph like this one would need to have the definition of an envy-free matching before defining the partition.  Sorry -- this is getting long. I hope I am on the right track.

Reply:
Thanks for the quick response!

I think it's best to stick to the original terms that were used in the article since we're basing everything on it. So $X_L$ and not $X_M$

The whole doc of the EFM function says: 

    Return a unique EFM (Envy Free Matching) partition of bipartite graph `G`.

    An EFM partition of a bipartite graph with bipartite sets X,Y is a unique partition of X - X_L, X_S and Y - Y_L, Y_S.
    This unique partition is proven to have the following properties:
    1. Every matching that matches all the nodes in X_L and is contained in G[X_L,Y_L] (all edges are between X_L and Y_L)
    is an envy-free matching.
    2. Every envy-free matching in `G` is contained in G[X_L,Y_L].

    Due to these properties we can say that X_L and Y_L are the "*good*" nodes and X_S and Y_S are the "*bad*" nodes,
    because - in the context of finding envy-free matchings, we will only find them in X_L and Y_L.

I thought the first sentence that appears when you mouse over should be short and people who want a better explanation can look in the docs to see the rest of it.

Do you think it's better that the whole description fits in the first description sentence?
I can also make some fixes in this description above, something like:

    Return a unique EFM (Envy Free Matching) partition of bipartite graph `G`.

    An EFM partition of a bipartite graph with bipartite sets X,Y is a unique partition of $ X := X_L U X_S $ and $ Y := Y_L U Y_S$.
    This unique partition is proven to have the following properties:
    1. Every matching that matches all the nodes in X_L and is contained in the induced subgraph G[X_L,Y_L] (all edges are between X_L and Y_L)
    is an envy-free matching.
    2. Every envy-free matching in `G` is contained in the induced subgraph G[X_L,Y_L].

    Due to these properties we can say that X_L and Y_L are the "*good*" nodes and X_S and Y_S are the "*bad*" nodes,
    because - in the context of finding envy-free matchings, we will only find them in X_L and Y_L.

Reply:
The first line should be short and can be less precise. It is used to describe a link to the docs for this function. That said, I would not use EFM or parens in it, just Envy Free Matching. We should define EFM in the paragraph.

I'm having trouble with the paragraph. It says that an EFM is a unique partition. And then it lists two sets that seem to be partitioned (so is it one partition or two partitions, or what?). Then it says that this partition is proven to have some properties -- but we haven't defined what makes the partition yet. How are we jumping to proving things about the partition when we haven't defined it yet.  Perhaps you mean that these properties **define** the partition and that it can be proven that this partition is unique??? but I'm guessing here.